lib/galaxy/datatypes/molecules.py

```diff
@@ -674,6 +795,637 @@ def set_peek(self, dataset):
+        visible=True,
+    )
+    MetadataElement(
+        name="chemical_formula",
+        desc="Chemical formula",
+        readonly=True,
+        visible=True,
+    )
+    MetadataElement(
+        name="is_periodic",
+        desc="Periodic boundary conditions",
+        readonly=True,
+        visible=True,
+    )
+    MetadataElement(
+        name="lattice_parameters",
+        desc="Lattice parameters",
+        readonly=True,
+        visible=True,
+    )
+
+    def sniff_prefix(self, file_prefix: FilePrefix):
+        """
+        Try to guess if the file is a CASTEP CELL file.
+
+        A fingerprint for CELL files is the use of %BLOCK and %ENDBLOCK to
+        denote data blocks (not case sensitive).
+
+        >>> from galaxy.datatypes.sniff import get_test_fname
+        >>> fname = get_test_fname('Si_uppercase.cell')
+        >>> Cell().sniff(fname)
+        True
+        >>> fname = get_test_fname('Si_lowercase.cell')
+        >>> Cell().sniff(fname)
+        True
+        >>> fname = get_test_fname('Si.cif')
+        >>> Cell().sniff(fname)
+        False
+        """
+        start_found = False
+        for line in file_prefix.line_iterator():
+            if not start_found and line.lower().startswith('%block'):
+                start_found = True
+            elif start_found and line.lower().startswith('%endblock'):
+                return True
+        return False
+
+    def set_meta(self, dataset, **kwd):
+        """
+        Find Atom IDs for metadata.
+        """
+        self.meta_error = False
+        if ase_io is None:
+            # Don't have optional dependency, can't set advanced values
+            return
+        else:
+            # enhanced metadata
+            try:
+                ase_data = ase_io.read(dataset.file_name, format="castep-cell")
+            except Exception as e:
+                log.warning('%s, set_meta Exception during ASE read: %s', self, unicodify(e))
+                self.meta_error = True
+                return
+
+            try:
+                atom_data = [
+                    str(sym) + str(pos) for sym, pos in zip(ase_data.get_chemical_symbols(), ase_data.get_positions())
+                ]
+                chemical_formula = ase_data.get_chemical_formula()
+                pbc = ase_data.get_pbc()
+                lattice_parameters = ase_data.get_cell().cellpar()
+            except Exception as e:
+                log.warning('%s, set_meta Exception during ASE metadata collection: %s', self, unicodify(e))
+                self.meta_error = True
+                return
+
+            # CELL file can only have one molecule
+            dataset.metadata.number_of_molecules = 1
+            dataset.metadata.atom_data = atom_data
+            dataset.metadata.number_of_atoms = len(dataset.metadata.atom_data)
+            dataset.metadata.chemical_formula = chemical_formula
+            try:
+                dataset.metadata.is_periodic = 1 if pbc else 0
```

Comment:
Would it make sense to replace 1 with True and 0 with False (everywhere you are setting `is_periodic`)? That would be more pythonic.

Reply:
I tried this originally, but it causes problems with JSON serialization: `TypeError: Object of type bool_ is not JSON serializable`

Full stacktrace (happens both for Cell which is a single bool and XYZ which is an array of bools)

```
galaxy.metadata DEBUG 2022-04-08 16:08:38,935 [pN:main,p:23712,tN:LocalRunner.work_thread-1] setting metadata externally failed for HistoryDatasetAssociation 1: Traceback (most recent call last):
  File "/home/qci88181/muon-galaxy/lib/galaxy/model/metadata.py", line 258, in to_JSON_dict
    encoded_meta_dict = galaxy.model.custom_types.json_encoder.encode(meta_dict)
  File "/usr/local/lib/python3.8/json/encoder.py", line 199, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/local/lib/python3.8/json/encoder.py", line 257, in iterencode
    return _iterencode(o, 0)
  File "/home/qci88181/muon-galaxy/lib/galaxy/model/custom_types.py", line 40, in default
    return json.JSONEncoder.default(self, obj)
  File "/usr/local/lib/python3.8/json/encoder.py", line 179, in default
    raise TypeError(f'Object of type {o.__class__.__name__} '
TypeError: Object of type bool_ is not JSON serializable

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/qci88181/muon-galaxy/lib/galaxy/metadata/set_metadata.py", line 407, in set_metadata_portable
    dataset.metadata.to_JSON_dict(filename_out)  # write out results of set_meta
  File "/home/qci88181/muon-galaxy/lib/galaxy/model/metadata.py", line 260, in to_JSON_dict
    raise Exception(f"Failed encoding metadata dictionary: {meta_dict}") from e
Exception: Failed encoding metadata dictionary: {'dbkey': '?', 'number_of_molecules': 1, 'atom_data': [['Si[0. 0. 0.]', 'Si[4.03575 4.03575 1.34525]', 'Si[2.6905 0.     2.6905]', 'Si[4.03575 1.34525 4.03575]', 'Si[0.     2.6905 2.6905]', 'Si[1.34525 1.34525 1.34525]', 'Si[1.34525 4.03575 4.03575]', 'Si[2.6905 2.6905 0.    ]']], 'number_of_atoms': [8], 'chemical_formula': ['Si8'], 'is_periodic': [False], 'lattice_parameters': [[0.0, 0.0, 0.0, 90.0, 90.0, 90.0]]}
```

Reply:
I've applied locally the 4 following suggested changes and uploading of the 3 test datasets worked for me:
```suggestion
                dataset.metadata.is_periodic = bool(pbc)
```

Reply:
Hmm, interesting. I'll commit them and try again locally.

Reply:
Huh, it worked. Not sure what I was doing wrong there.